Address derivation scheme as type parameter to WalletLayer

rvl opened the pull request:
Relates to issue #557 

# Overview

This adds a key type parameter to the `WalletLayer`. The key type must satisfy `WalletKey`.

The database layer is parameterised by key type. For sqlite, the type must satisfy `PersistKey`.

There are instances of the key for new types `SeqKey` and `RndKey`.

For now, the API server and many other things hardcode the key parameter to `SeqKey`.

# Comments

I worked back from the `keyToAddress` function. The most complicated combination is a random scheme old-style address for Testnet.

Going from that I felt that it was necessary to include the target (e.g. `HttpBridge 'Testnet`) as a second parameter of the `KeyToAddress` typeclass.

```
class KeyToAddress target (key :: Depth -> * -> *) where
    keyToAddress :: key 'AddressK XPub -> Address
```

The `rndToAddress` function is stubbed out and can be implemented in a future PR.

## Things to do
- [x] Delete `Key` newtype and don't wrap it in `SeqKey` and `RndKey` wrap 
- [x] Remove the `unwrapKey` function. It's just a temporary helper.
- [x] Add key parameter to `IsOwned` class.
- [x] Integration tests building.
- [x] Figure out what fields to be in `RndKey`
- [x] implement `instance WalletKey RndKey`
- [x] Shift a few functions from `AddressDerivation` to `AddressDerivation.Sequential`
- [x] The restore benchmark doesn't build, and the error is a little tricky.
- [x] Rework from reviews 7/7/2019
- [ ] Git rebase and squash

## Things to do in next PR
- [ ] Remove hardcoding of `SeqKey` as much as possible, to make things more polymorphic.
- [ ] Implement `instance KeyToAddress t RndKey`
- [x] Fix type signature of `Random.encodeDerivationPath` because derivation path is now included in the `RndKey`.
- [ ] Implement `instance PersistKey RndKey`
- [ ] Remove `AddressDerivation.Common` module (actually maybe we can keep it for common functions used by `PersistKey`).

rvl committed:
Address derivation scheme as type parameter to WalletLayer

paweljakubas commented on lib/core/src/Cardano/Wallet/Primitive/AddressDerivation/Random.hs:
```diff
@@ -77,6 +84,22 @@ import qualified Data.ByteArray as BA
+-- | Data for generating random scheme addresses
+data RndKey (level :: Depth) key = RndKey
+   { rootKey :: Key 'RootK XPrv
+   , addrKey :: Key 'AddressK XPub -- address public key
```
why public if above root private key?

rvl replied:
Yes, it was quite wrong.

paweljakubas commented on lib/http-bridge/src/Cardano/Wallet/HttpBridge/Compatibility.hs:
```diff
@@ -111,6 +125,32 @@ instance KeyToAddress (HttpBridge 'Testnet) where
+
+rndToAddressAttrs, rndToAddressAttrsWithProtocolMagic
+    :: Key 'AddressK XPub
+    -> Key 'RootK XPub
```
just nitpick, but maybe we want `Key 'RootK XPub` to be first argument ...

rvl replied:
True, changed.

paweljakubas commented on default.nix:
```diff
@@ -46,6 +46,7 @@ in {
       cardano-wallet-core-integration
       cardano-wallet-http-bridge
       cardano-wallet-jormungandr
+      cardano-wallet-shelley
```
why it is needed?

rvl replied:
Lets me use ghci. I can move it to another PR.

KtorZ commented on lib/core/src/Cardano/Wallet/Primitive/AddressDerivation/Random.hs:
```diff
@@ -77,6 +89,33 @@ import qualified Data.ByteArray as BA
+-- | Data for generating random scheme addresses
+data RndKey (level :: Depth) key = RndKey
+    { rootKey :: Key 'RootK XPrv
+    , addrKey :: Key 'AddressK XPub -- address public key
```
Shouldn't this second field be a `Key level key` :thinking: ?
We only have an address key for the `AddressK` depth, and it can be public or private. So in practice we do manipulate:

```
RndKey 'RootK XPrv
RndKey 'RootK XPub
RndKey 'AccountK XPrv
RndKey 'AddressK XPrv
RndKey 'AddressK XPub
```

We do always have a root key, but we only have an account index for keys below the `AccountK` depth, and only an address index for keys at `AddressK` depth.

paweljakubas replied:
If I understand it properly, we need 
`Key 'RootK XPrv` (or alternatively seed and encryption passwords) AND
`encryption password`, `account index` and `address index`.
With those we can derive key for any address. This is done (1) : 
```
rootXPrv = generateKeyFromSeed seed encPwd
```
in order to get root private key that of type `Key 'RootK XPrv`
Knowing that we can get (2) : 
```   
accXPrv = deriveAccountPrivateKey encPwd rootXPrv accIx
addrXPrv = deriveAddressPrivateKey encPwd accXPrv addrIx
```
or public counterparts by applying `publicKey` to any private key.
Saying that can't `RndKey` be : 
```   
data RndKey (level :: Depth) key = RndKey
    { rootKey :: Key 'RootK XPrv
    , encryptionPassword :: Password "encryption" 
    , accountIndex :: Index 'Hardened 'AccountK
    , addressIndex :: Index 'Soft 'AddressK
    } deriving stock (Generic)
```
?

KtorZ replied:
@paweljakubas key derivation is a rather expensive operation that ideally we don't want to be re-performing all the time so we would need to at least store the cryptographic content inside the `RndKey` (what we currently wrap in a `Key (* :: Depth) *`. 

Having said that, note that we either need the root public key, or the already computed payload passphrase (which is a different thing than the master encryption passphrase!) which is specific to the rnd scheme and can be computed from the root public key. 

NOTE:
The reason we want to store derivation indexes and the payload encryption passphrase inside the key is to make key self-sufficient to compute addresses from keys. Because in order to compute addresses in the legacy format we need these pieces of information, we embed them with the cryptographic materials and make the manipulation of keys rather similar between both (seq and rnd) schemes.

rvl replied:
Thanks for the help with this guys. Please check out the new version of `RndKey`. It does now contain enough information to implement `rndToAddress`. But it's sort of weird because it has a `level` type parameter which is not used in any of the fields.

paweljakubas commented on lib/core/src/Cardano/Wallet/Primitive/AddressDerivation/Random.hs:
```diff
@@ -314,12 +394,12 @@ decodeDerPath = decodeListIndef CBOR.decodeWord32 >>= \case
     (PBKDF2.prfHMAC SHA512)
     (PBKDF2.Parameters 500 32)
-    (unXPub rootXPub)
+    (unXPub . toXPub . getKey $ masterKey)
```
why to use `toXPub` if we have : 
```
publicKey
    :: Key level XPrv
    -> Key level XPub
publicKey (Key k) =
    Key (toXPub k)
```
in `Cardano.Wallet.Primitive.AddressDerivation` ?

paweljakubas replied:
ok, I have only now detected that `publicKey` does not take form like this anymore

paweljakubas commented on lib/core/src/Cardano/Wallet/Primitive/AddressDerivation/Random.hs:
```diff
@@ -369,3 +449,46 @@ decodeNestedBytes dec bytes =
+    :: RndKey level XPrv
+    -> RndKey level XPub
+publicKeyRnd k = k
+    { masterKey = toXPub <$> masterKey k
```
the same remark about `toXPub` as above

paweljakubas replied:
ok, I have only now detected that `publicKey` does not take form like this anymore

paweljakubas commented on lib/core/src/Cardano/Wallet/Primitive/AddressDerivation/Sequential.hs:
```diff
@@ -197,30 +242,161 @@ deriveAddressPrivateKey
+-------------------------------------------------------------------------------}
+
+-- | Extract the public key part of a private key.
+publicKeySeq
```
I would reuse 
```
publicKey
    :: Key level XPrv
    -> Key level XPub
```
here

paweljakubas replied:
ok, I have only now detected that `publicKey` does not take form like this anymore

paweljakubas commented on lib/core/test/unit/Cardano/Wallet/Primitive/AddressDerivation/RandomSpec.hs:
```diff
@@ -207,8 +208,8 @@ decodeTest DecodeDerivationPath{..} =
-    Key rootXPrv = generateKeyFromSeed (Passphrase seed) (Passphrase "")
-    rootXPub = Key (toXPub rootXPrv)
+    rootXPrv = generateKeyFromSeed (Passphrase seed) (Passphrase "") :: RndKey 'RootK XPrv
+    rootXPub = publicKey rootXPrv
```
yeah, here you use `publicKey` :+1:

paweljakubas committed:
Benchmark fixes and review rework

* fix bench_restore

* use type parameter rather than exact type

KtorZ commented on exe/wallet/jormungandr/Main.hs:
```diff
@@ -345,7 +347,7 @@ cmdServe = command "serve" $ info (helper <*> cmd) $ mempty
             :: CM.Configuration
             -> Trace IO Text
-            -> (DBLayer IO s t -> IO a)
+            -> (DBLayer IO s t SeqKey -> IO a)
```
Minor note but, this should / could be `k`, as aliased above :point_up:

paweljakubas replied:
fixed in https://github.com/input-output-hk/cardano-wallet/pull/597/commits/d7577071415476ffcd23e89749d9391e37bff24c

KtorZ commented on exe/wallet/http-bridge/Main.hs:
```diff
@@ -240,7 +241,7 @@ cmdServe = command "serve" $ info (helper <*> cmd) $ mempty
     exec
         :: forall t n s. (t ~ HttpBridge n, s ~ SeqState t)
-        => (KeyToAddress t, KnownNetwork n)
+        => (KeyToAddress t SeqKey, KnownNetwork n)
```
could use a type-alias `k ~ SeqKey` for consistency here and below too.

paweljakubas replied:
addressed in https://github.com/input-output-hk/cardano-wallet/pull/597/commits/72bc51d1290deb86a7db638be34de9edc206e296

KtorZ commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -313,16 +315,17 @@ getWalletWithCreationTime w (ApiT wid) = do
-    => WalletLayer (SeqState t) t
+    :: forall t key.
+       (DefineTx t, KeyToAddress t key, key ~ SeqKey)
+    => WalletLayer (SeqState t) t key
```
Note: parameter is somtimes called `key` and sometimes `k`. It'll help reading and consistency to use a single name everywhere.

paweljakubas replied:
addressed in https://github.com/input-output-hk/cardano-wallet/pull/597/commits/72bc51d1290deb86a7db638be34de9edc206e296

paweljakubas commented on lib/core/src/Cardano/Wallet/Primitive/AddressDerivation.hs:
```diff
@@ -164,21 +164,6 @@ instance Enum (Index 'Soft level) where
 data DerivationType = Hardened | Soft
 
--- | Extract the public key part of a private key.
-publicKey
```
ok, now I got it `publicKey` now is method of `WalletKey`

KtorZ commented on lib/core/src/Cardano/Wallet/Primitive/AddressDerivation.hs:
```diff
@@ -394,102 +365,44 @@ instance MonadRandom ((->) (Passphrase "salt")) where
+        :: Passphrase "encryption-old"
+        -> Passphrase "encryption-new"
+        -> key level XPrv
+        -> key level XPrv
```
~~level~~ ->  `depth`, for consistency

paweljakubas replied:
addressed in https://github.com/input-output-hk/cardano-wallet/pull/597/commits/8ea1f7ab680282110c755ce34461da5b0dd231c1

KtorZ commented on lib/core/src/Cardano/Wallet/Primitive/AddressDerivation.hs:
```diff
@@ -394,102 +365,44 @@ instance MonadRandom ((->) (Passphrase "salt")) where
+    -- unwrapKey :: key level depth -> Key level depth
+
+    publicKey :: key level XPrv -> key level XPub
+    digest :: HashAlgorithm a => key level XPub -> Digest a
```
Same here, ~~level~~ -> `depth`

paweljakubas replied:
addressed in https://github.com/input-output-hk/cardano-wallet/pull/597/commits/8ea1f7ab680282110c755ce34461da5b0dd231c1

KtorZ commented on lib/core/src/Cardano/Wallet/Primitive/AddressDerivation/Common.hs:
```diff
@@ -15,43 +14,28 @@
--- left out of the sum type.
-data Depth = RootK | AccountK | AddressK
+toHexText :: ByteString -> ByteString
+toHexText = convertToBase Base16
```
I wonder how useful really are these helpers which we probably won't use consistently in the codebase :/ 
We are still one import away from this, and that's something which feels unrelated to address discovery, so, feels awkward to have in this location.

I guess they are simply leftovers of moving parts away. But we could find a better home for them (and maybe better names, toHex**Text** actually works with **ByteString**, which is kinda misleading :sweat_smile:  ...

rvl replied:
Let's wait until `instance PersistKey RndKey` is implemented before moving them.

KtorZ commented on lib/core/src/Cardano/Wallet/Primitive/AddressDerivation/Random.hs:
```diff
@@ -77,6 +94,50 @@ import qualified Data.ByteArray as BA
+
+-- | Data for generating random scheme addresses
+data RndKey (level :: Depth) key = RndKey
+    { masterKey :: Key 'RootK key
```
I am not sure the `Key` wrapper is buying us much? The business logic will already manipulate keys only through the WalletKey interface, so, having another level of indirection within the abstraction boundary feels like overhead. Although, it can help disambiguate the root key with the rest perhaps.

Not very strong about this but I think that using `key` directly would be sufficient at this level.

rvl replied:
The only place it _might_ help is `hdPassphrase`, but I agree a second type wrapper is not necessary.

KtorZ commented on lib/core/src/Cardano/Wallet/Primitive/AddressDerivation/Random.hs:
```diff
@@ -77,6 +94,50 @@ import qualified Data.ByteArray as BA
+    -- ^ The address key, generated from the account key and address index.
+    , accountIndex :: Index 'Hardened 'AccountK
+    -- ^ Which account in the hierarchical scheme to generate addresses in.
+    , addressIndex :: Index 'Soft 'AddressK
```
I am failing to see how we can generate a random root key and fill the account and address key fields? (same for the indexes). 
I think there's still some confusion / disagreement about what's the place of a `Key` in our codebase. Now the `RndKey` looks like a fully instantiated address key. We've lost the ability to represent keys at different steps in the hierarchy.

KtorZ replied:
A `RndKey` needs to capture a few things:

- The current depth in the hierarchy (Root, Account, Address)
- Whether the key is private or public
- The current derivation path 

So, a possible structure for this would be:

```hs
data RndKey (depth :: Depth) key = RndKey
    { rawKey :: key
    , derivationPath :: DerivationPath depth
    }

type family DerivationPath (depth :: Depth) :: * where
    DerivationPath 'RootK =
        ()
    DerivationPath 'AccountK =
        Index 'Hardened 'AccountK
    DerivationPath 'AddressK =
        (Index 'Hardened 'AccountK, Index 'Hardened 'AccountK)
```

This allows us to represent non-ambiguously the various type of keys we need for the random derivation, without the need for placeholder data.
Now, in order to implement `KeyToAddress`, we need an extra bit of information which is the actual passphrase derived from the public root key associated to a particular address private key. 

I see three options:

1/ Modify `KeyToAddress` to also require an extra argument. However, this need to be an associated type family from the type class since sequential key doesn't require any extra argument. So, something in the lines of:

```hs
class KeyToAddress target key where
    type KeyToAddressArg key :: *
    keyToAddress :: KeyToAddressArg key -> key 'AddressK XPub -> Address

instance KeyToAddress t SeqKey where
     type KeyToAddressArg SeqKey = ()
     ...

instance KeyToAddress t RndKey where
    type KeyToAddressArg RndKey = RndKey 'RootK XPub
    ...
```

2/ Or, we can directly embed the root public key in any `RndKey` we have, and extend the definition with an extra `rootKey` field.

```hs
data RndKey (depth :: Depth) key = RndKey
    { rawKey :: key
    , derivationPath :: DerivationPath depth
    , rootKey :: key -- For derivation payload encryption 
    }
```

or, alternatively (preferred):

3/ Store only the encryption passphrase with the key (i.e., derive it when we generate the key from the seed initially, and keep it there): 

```hs
data RndKey (depth :: Depth) key = RndKey
    { rawKey :: key
    , derivationPath :: DerivationPath depth
    , payloadPassphrase :: Passphrase "address-derivation-payload"
    }
```

paweljakubas replied:
thanks for this explanation!

rvl replied:
Thanks, a top-level type synonym family for the derivation path makes it work.

KtorZ commented on lib/core/src/Cardano/Wallet/Primitive/AddressDerivation/Random.hs:
```diff
@@ -89,8 +150,8 @@ deriveAccountPrivateKey
     -> Key 'RootK XPrv
     -> Index 'Hardened 'AccountK
     -> Key 'AccountK XPrv
```
I'd rather have the signature be:

```hs
deriveAccountPrivateKey
    :: Passphrase "encryption"
    -> RndKey 'RooK XPrv
    -> Index 'Hardened 'AccountK
    -> RndKey 'AccountK XPrv
```

That's actually at this moment we know what account derivation index is used. (See previous remark about the `RndKey` structure).

KtorZ commented on lib/core/src/Cardano/Wallet/Primitive/AddressDerivation/Random.hs:
```diff
@@ -104,26 +165,45 @@ deriveAddressPrivateKey
+    accountIndex = Index idx
+    addressIndex = Index idx
+    -- The initial index of both the account and address.
+    idx = 2147483648
```
Hmmm.. This is rather confusing: address index and account index are pre-filled with a dummy value, though they are not known at this stage (when generating a root key from the seed). Plus, I don't see them being updated anywhere in the derivation functions. 

Also, side-note but, avoid hard-coded magic values like `2147483648`. `Index` has a `Bounded` instance, so we can go for `minBound @(Index 'AccountK 'Hardened)` and `minBound @(Index 'AddressK 'Hardened)`

KtorZ commented on lib/core/src/Cardano/Wallet/Primitive/AddressDerivation/Random.hs:
```diff
@@ -104,26 +165,45 @@ deriveAddressPrivateKey
     -> Key 'AccountK XPrv
     -> Index 'Soft 'AddressK
```
I think I made a mistake when writing down the ticket but address derivation index is also hardened in the legacy scheme.

KtorZ commented on lib/core/src/Cardano/Wallet/Primitive/AddressDerivation/Random.hs:
```diff
@@ -77,6 +94,50 @@ import qualified Data.ByteArray as BA
+
+instance WalletKey RndKey where
+    type WalletKeySeed RndKey = Passphrase "seed"
+    unsafeGenerateKeyFromSeed = unsafeGenerateKeyFromSeedRnd
```
I don't think `(unsafe)GenerateKeyFromSeed` should be part of the abstraction here. This function acts as a constructor in my opinion, which is the way by which we get to instantiate a particular `RndKey` or `SeqKey`. Therefore, they might have different arguments which is fine because we never actually call this from the wallet layer. This is done outside of the wallet logic.

rvl replied:
I pulled the generate functions out of the class. But actually, whether to put this in the type class might depend on how we generalise the `Server.postWallet` function.

KtorZ replied:
@rvl We'll have different endpoints for creating wallets using sequential and random scheme, because they have fundamentally different parameters. 
So, in practice, this is where we make the concrete choice for the key type (and wallet layer).

KtorZ reviewed (changes requested):
There's a confusion with what the sort of information captured in a `RndKey` which we need to resolve before moving on.

KtorZ committed:
Make key a type-parameter of 'TransactionLayer'

KtorZ commented on lib/http-bridge/test/bench/Main.hs:
```diff
@@ -199,26 +196,25 @@ parseNetwork = \case
+    => (WalletId, WalletName, s)
     -> IO ()
-bench_restoration _ _ (wid, wname, s) = withHttpBridge network $ \port -> do
+bench_restoration (wid, wname, s) = withHttpBridge network $ \port -> do
```
NOTE: I've removed the proxies in favor of `AllowAmbiguousTypes` + `TypeApplications`. This is equivalent but I find the syntax with type applications much clearer.

paweljakubas committed:
Final compilation fixes and review rework

* final compilation fix

* key -> k in forall and complying http-bridge exe

* fix naming consistency in digest and changePassphrase

* small fix

rvl committed:
Rework RndKey

rvl committed:
Move WalletKey generation functions out of type class

rvl committed:
Inline some class instance functions and fix changePassphraseRnd

rvl committed:
Rework "instance KeyToAddress t RndKey" stubs

rvl committed:
Make it compile

rvl committed:
More making it compile

rvl committed:
Do encodeDerivationPath :: RndKey 'AddressK XPub -> CBOR.Encoding

rvl committed:
Fix the haddocks

rvl committed:
A little bit more stubbing of rndKeyToAddress

rvl committed:
stylish-haskell wanted FlexibleContexts

And hlint doesn't object to that.

KtorZ committed:
remaining consistency and minor tweaks

- Move class-constraints around
- Implement 'getRawKey' for RndKey
- Review a few type signatures
- Remove unneeded comments

paweljakubas reviewed (approved):
LGTM now!
